### PR TITLE
Add onnx support for `torch.tensor_split` 

### LINKED ISSUE #73454: Support tensor_split in ONNX export
### 🚀 The feature, motivation and pitch

When working with BERT models, doing tensor splits is very common practice. However, such operation is currently not supported in ONNX export. 

When splitting using a list of integers, the tracing does not work if the list is of variable length. When using torhc.tensor_split, the export does not work as it is currently not supported.

Would be great if we can have this supported in ONNX export. Thanks!

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by qqaatw**
Fixes #73454

### CODE REVIEW

**Comment by justinchuby** [comment]
Thanks for your contribution!

File: torch/onnx/symbolic_opset13.py [hunk 1]
```diff
@@ -109,6 +109,136 @@ def unsafe_split_with_sizes(g, self, split_sizes, dim, _outputs=None):
...
+
+            return g.op("SequenceInsert", loop_out, last_slice)
+
+        else:  # scalar tensor
+            dim_size = symbolic_helper._size_helper(g, self, axis)
+            min_split_size = g.op("Div", dim_size, indices_or_sections)
+            min_split_size_plus_1 = g.op(
+                "Add",
+                min_split_size,
+                g.op("Constant", value_t=torch.tensor(1, dtype=torch.long)),
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:196]: For readability move `const_1` outside the if block and use it here too.

File: torch/onnx/symbolic_opset13.py [hunk 2]
```diff
@@ -109,6 +109,136 @@ def unsafe_split_with_sizes(g, self, split_sizes, dim, _outputs=None):
     return split_with_sizes(g, self, split_sizes, dim, _outputs)
 
 
+@symbolic_helper.parse_args("v", "v", "i", "i")
+def tensor_split(g, self, indices_or_sections, dim, _outputs=None):
+    axis = g.op("Constant", value_t=torch.tensor(dim, dtype=torch.long))
+    axis = opset11.unsqueeze(g, axis, 0)
+
+    def static_tensor_split(self, indices):
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:117]: since this is only used in one place, just inline the code rather than defining a helper function

File: torch/onnx/symbolic_opset13.py [hunk 3]
```diff
@@ -109,6 +109,136 @@ def unsafe_split_with_sizes(g, self, split_sizes, dim, _outputs=None):
...
+        res.append(g.op("Slice", self, start, end, axis))
+        return res
+
+    if not symbolic_helper._is_split_static(indices_or_sections, _outputs):
+        is_scalar = symbolic_helper._get_tensor_rank(indices_or_sections) == 0
+
+        if (
+            symbolic_helper._is_tensor(indices_or_sections) and not is_scalar
+        ):  # 1-d tensor
+            # loop conditions
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:135]: This comment is not adding anything beyond what the variable names already make clear. Please remove it.

File: torch/onnx/symbolic_opset13.py [hunk 4]
```diff
@@ -109,6 +109,136 @@ def unsafe_split_with_sizes(g, self, split_sizes, dim, _outputs=None):
...
+            )
+            loop_len = opset11.unsqueeze(g, loop_len, 0)
+            loop_condition = g.op("Cast", const_1, to_i=9)
+
+            padding_0 = g.op("Constant", value_t=torch.tensor([0], dtype=torch.long))
+            indices_or_sections = g.op(
+                "Concat", padding_0, indices_or_sections, axis_i=0
+            )
+
+            # Create an empty sequence to store final expansions
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:148]: ditto on comment not adding value, please remove

File: torch/onnx/symbolic_opset13.py [hunk 5]
```diff
@@ -109,6 +109,136 @@ def unsafe_split_with_sizes(g, self, split_sizes, dim, _outputs=None):
...
+        ):  # 1-d tensor
+            # loop conditions
+            const_1 = g.op("Constant", value_t=torch.tensor(1, dtype=torch.long))
+            loop_len = symbolic_helper._size_helper(
+                g, indices_or_sections, g.op("Constant", value_t=torch.tensor(0))
+            )
+            loop_len = opset11.unsqueeze(g, loop_len, 0)
+            loop_condition = g.op("Cast", const_1, to_i=9)
+
+            padding_0 = g.op("Constant", value_t=torch.tensor([0], dtype=torch.long))
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:143]: please add a comment explaining why padding is needed

File: torch/onnx/symbolic_opset13.py [hunk 6]
```diff
@@ -109,6 +109,136 @@ def unsafe_split_with_sizes(g, self, split_sizes, dim, _outputs=None):
...
+                indices_or_sections,
+                loop_block.op("Add", block_input_iter, const_1),
+                axis_i=0,
+            )
+
+            slice = loop_block.op("Slice", self, start, end, axis)
+            final_splits = loop_block.op("SequenceInsert", final_splits, slice)
+
+            # Loop outputs
+            cond_out = loop_block.op("Cast", loop_condition, to_i=9)
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:172]: why is this cast needed? loop_condition was already cast to bool on line 141

File: torch/onnx/symbolic_opset13.py [hunk 7]
```diff
@@ -109,6 +109,136 @@ def unsafe_split_with_sizes(g, self, split_sizes, dim, _outputs=None):
...
+    split_val = indices_or_sections.node()["value"]
+    if split_val.dim() > 0:
+        return static_tensor_split(self, indices_or_sections)
+    split_size = symbolic_helper._get_const(
+        indices_or_sections, "i", "indices_or_sections"
+    )
+
+    size = symbolic_helper._get_tensor_dim_size(self, dim)
+    if size is None:
+        if _outputs is not None:
```
Comment (garymm) [inline torch/onnx/symbolic_opset13.py:224]: It doesn't make sense for `_outputs == 0`, right?
If so, then make this more pythonic with:
```suggestion
        if _outputs:
```

**Review by garymm — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: torch/onnx/symbolic_opset13.py (hunk 6 above)
Comment (qqaatw) [reply torch/onnx/symbolic_opset13.py]: No need indeed. I changed it to `Identity`.

File: torch/onnx/symbolic_opset13.py (hunk 7 above)
Comment (qqaatw) [reply torch/onnx/symbolic_opset13.py:224]: Nearly all of the other `_outputs` checks in `symbolic_opsetX.py` files compare with an integer or `None`, so I think we just leave it as is (maybe more explicit).

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 8]
```diff
@@ -6125,6 +6125,56 @@ def forward(self, x):
...
+                return input.tensor_split([2, 3, 5])
+
+        x = torch.randn(5, 4, 3)
+        self.run_test(TensorSplitModel3(), x)
+
+    @skipIfUnsupportedMinOpsetVersion(13)
+    def test_tensor_split_scalar(self):
+        class TensorSplitModel(torch.nn.Module):
+            def forward(self, x):
+                return torch.tensor_split(x, x.size(1))
```
[the remark this answers is not in the document]
Comment (qqaatw) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:6150]: In fact, all the tests presented trigger the dynamic code, I think there are only some edge cases that will trigger the static code, and I'm not aware of how to make a PyTorch tensor or Python scalar that will represent as constant after converted to Torch JiT or ONNX graph. 

Here is the condition, which is also used in the onnx conversion of `torch.split` in the same file:
```python
def _is_split_static(split_size_or_sizes, _outputs):
    if _outputs is None:
        return False
    if (
        _is_value(split_size_or_sizes)
        and split_size_or_sizes.node().kind() != "onnx::Constant" # the type of the value should be `onnx::Constant`
    ):
        return False
    return True
```

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 9]
```diff
@@ -6125,6 +6125,56 @@ def forward(self, x):
...
+
+        x = torch.randn(1, 2, 3, requires_grad=True)
```
[the remark this answers is not in the document]
Comment (qqaatw) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:6152]: The `requires_grad` is checked here when run `self.run_test`:

https://github.com/pytorch/pytorch/blob/df748b60f7bfc0a6b031ad151a0b44f9d68ff52b/torch/onnx/verification.py#L33-L48

File: torch/onnx/symbolic_opset13.py [hunk 10]
```diff
@@ -109,6 +109,136 @@ def unsafe_split_with_sizes(g, self, split_sizes, dim, _outputs=None):
...
+        for i in range(_outputs - 1):
+            end = g.op("Gather", indices_or_sections, i, axis)
+            res.append(g.op("Slice", self, start, end, axis))
+            start = end
+
+        end = symbolic_helper._size_helper(g, self, axis)
+        res.append(g.op("Slice", self, start, end, axis))
+        return res
+
+    if not symbolic_helper._is_split_static(indices_or_sections, _outputs):
```
[the remark this answers is not in the document]
Comment (qqaatw) [reply torch/onnx/symbolic_opset13.py:118]: Since in most scenarios the dynamic code will be triggered, would it be better to put the dynamic code in front of the static one?

The conversion of `torch.split` in this file adopts the same approach.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (qqaatw) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:6152]: I'm ok with removing it.

File: torch/onnx/symbolic_opset13.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (garymm) [reply torch/onnx/symbolic_opset13.py:118]: Well if we can't find any input that triggers the static case, I think we should remove the static case altogether.
But if we do keep both, the organization of the code should be optimized for readability, not based on how often we guess each branch of the if will be executed.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (garymm) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: Remove it, it doesn't make any difference except making the code slower.

File: torch/onnx/symbolic_opset13.py (hunk 7 above)
Comment (garymm) [reply torch/onnx/symbolic_opset13.py:224]: OK, makes sense to keep it consistent for now.

**Review by garymm — CHANGES_REQUESTED** [review]

File: torch/onnx/symbolic_opset13.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (qqaatw) [reply torch/onnx/symbolic_opset13.py]: OK, makes sense.

File: torch/onnx/symbolic_opset13.py [hunk 11]
```diff
@@ -115,7 +115,48 @@ def tensor_split(g, self, indices_or_sections, dim, _outputs=None):
...
+        num_splits_one_extra = size % split_size
+
+        splits = num_splits_one_extra * [min_split_size + 1]
+        leftover = (split_size - num_splits_one_extra) * [min_split_size]
+
+        splits = g.op(
+            "Constant", value_t=torch.tensor(splits + leftover, dtype=torch.long)
+        )
+        return g.op("Split", self, splits, axis_i=dim, outputs=_outputs)
+    else:
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset13.py:159]: I think you can remove the else and de-indent the rest of the logic

**Review by garymm — APPROVED** [review]
Thanks!
Please verify CI checks pass, then you can ask the pytorchbot to merge.
